neqo-http3/src/connection_client.rs

```diff
@@ -5358,4 +5368,64 @@ mod tests {
             .recv_if_encoder_stream(&mut server.conn, CLIENT_SIDE_DECODER_STREAM_ID);
         assert_eq!(server.encoder.stats().stream_cancelled_recv, 0);
     }
+
+    #[test]
+    fn multiple_streams_in_decoder_blocked_state() {
+        let (mut client, mut server, request_stream_id) = connect_and_send_request(true);
+
+        setup_server_side_encoder(&mut client, &mut server);
+
+        let headers = vec![
+            (String::from(":status"), String::from("200")),
+            (String::from("my-header"), String::from("my-header")),
+            (String::from("content-length"), String::from("0")),
+        ];
+        let encoded_headers = server
+            .encoder
+            .encode_header_block(&mut server.conn, &headers, request_stream_id)
+            .unwrap();
+        let hframe = HFrame::Headers {
+            header_block: encoded_headers.to_vec(),
+        };
+
+        // Delay encoder instruction so that the stream will be blocked.
+        let encoder_insts = server.conn.process(None, now());
+
+        // Send response headers.
+        let mut d = Encoder::default();
+        hframe.encode(&mut d);
+        server_send_response_and_exchange_packet(
+            &mut client,
+            &mut server,
+            request_stream_id,
+            &d,
+            true,
+        );
+
+        // Headers are blocked waiting fro the encoder instructions.
+        let header_ready_event = |e| matches!(e, Http3ClientEvent::HeaderReady { .. });
+        assert!(!client.events().any(header_ready_event));
+
+        // Make another request.
+        let request2 = make_request_and_exchange_pkts(&mut client, &mut server, true);
+        // Send response headers.
+        server_send_response_and_exchange_packet(&mut client, &mut server, request2, &d, true);
+
+        // Headers on the second request are blocked as well are blocked
+        // waiting fro the encoder instructions.
+        assert!(!client.events().any(header_ready_event));
+
+        // Now read headers.
+        let _ = client.process(encoder_insts.dgram(), now());
+
+        // find the new request/response stream and send frame v on it.
+        let mut count_responses = 0;
+        while let Some(e) = client.next_event() {
```

Comment:
I've been wondering whether we should provide an `Iterator` implementation for events so that we can do these things more efficiently.  The code you have here is roughly:

```rust
client.events().filter(|e| matches!(e, Http3ClientEvent::HeaderReady { .. })).count();
```

We could get that by doing:

```rust
trait ConnectionEventProvider {
    fn next_event(&mut self) -> Option<ConnectionEvent>;
}

struct<'a, T> ConnectionEventIter<T> {
    c: &'a T;
}
    
impl<'a, T> Iterator for ConnectionEventIter<'a, T> 
    where T: ConnectionEventProvider {
    type Item = ConnectionEvent;
    fn next(&mut self) -> Option<Self::Item> {
        self.c.next_event()
    }
}
```